rpc/src/client/transport/router.rs

```diff
@@ -0,0 +1,182 @@
+//! Event routing for subscriptions.
+
+use crate::client::subscription::SubscriptionTx;
+use crate::event::Event;
+use std::collections::{HashMap, HashSet};
+use tracing::debug;
+
+/// Provides a mechanism for tracking [`Subscription`]s and routing [`Event`]s
+/// to those subscriptions.
+///
+/// [`Subscription`]: struct.Subscription.html
+/// [`Event`]: ./event/struct.Event.html
+#[derive(Debug)]
+pub struct SubscriptionRouter {
+    // A map of subscription queries to collections of subscription IDs and
+    // their result channels. Used for publishing events relating to a specific
+    // query.
+    subscriptions: HashMap<String, HashMap<String, SubscriptionTx>>,
+}
+
+impl SubscriptionRouter {
+    /// Publishes the given event to all of the subscriptions to which the
+    /// event is relevant. At present, it matches purely based on the query
+    /// associated with the event, and only queries that exactly match that of
+    /// the event's.
+    pub fn publish(&mut self, ev: &Event) -> PublishResult {
+        let subs_for_query = match self.subscriptions.get_mut(&ev.query) {
+            Some(s) => s,
+            None => return PublishResult::NoSubscribers,
+        };
+        // We assume here that any failure to publish an event is an indication
+        // that the receiver end of the channel has been dropped, which allows
+        // us to safely stop tracking the subscription.
+        let mut disconnected = HashSet::new();
+        for (id, event_tx) in subs_for_query {
+            if let Err(e) = event_tx.send(Ok(ev.clone())) {
+                disconnected.insert(id.clone());
+                debug!(
+                    "Automatically disconnecting subscription with ID {} for query \"{}\" due to failure to publish to it: {}",
+                    id, ev.query, e
+                );
+            }
+        }
+        // Obtain a mutable reference because the previous reference was
```

Comment:
interesting 🤔 i would expect Rust to figure out that the loop is over

Reply:
`for x in y { b }` is syntactic sugar for `let mut iter = y.into_iter(); while let Some(x) = iter.next() { b }` and [`into_iter()`](https://doc.rust-lang.org/stable/std/iter/trait.IntoIterator.html#tymethod.into_iter) consumes `self`, therefore one cannot access the collection being iterated over (ie. `subs_for_query`) after the loop is over.

On the other hand, `for x in &y { b }` would desugar to a call to `iter()` which does not not consume `self` but instead yields reference to the underlying elements. Same with `for x in &mut y`.

I wonder though if we could use `for (id, event_tx) in &subs_for_query` here instead?

Reply:
got it! thanks for explaining it @romac 🙏

Reply:
So interestingly, the `for (id, event_tx) in &subs_for_query` didn't work previously, and I've checked again now, it results in the following error:

```
error[E0277]: `&&mut std::collections::HashMap<std::string::String, client::sync::ChannelTx<std::result::Result<event::Event, error::Error>>>` is not an iterator
  --> rpc/src/client/transport/router.rs:35:31
   |
35 |         for (id, event_tx) in &subs_for_query {
   |                               ^^^^^^^^^^^^^^^ `&&mut std::collections::HashMap<std::string::String, client::sync::ChannelTx<std::result::Result<event::Event, error::Error>>>` is not an iterator
   |
   = help: the trait `std::iter::Iterator` is not implemented for `&&mut std::collections::HashMap<std::string::String, client::sync::ChannelTx<std::result::Result<event::Event, error::Error>>>`
   = note: required by `std::iter::IntoIterator::into_iter`
```

`for (id, event_tx) in &mut subs_for_query` also doesn't compile.

The following code, however, does work, so I'll update the code accordingly:

```rust
for (id, event_tx) in subs_for_query.borrow_mut() {
    // ...
}
```

I feel as though if I manually expanded all the syntactic sugar, I'd understand this better 😁